### PR TITLE
Helper class to test scheme functionality on all elements by DFS

### PR DESCRIPTION
**Pull request by lukasdreyer**
**_Describe your changes here:_**

Helper class to test scheme functionality on all elements by DFS

Also enables this for `find_parent`, more to come

**_All these boxes must be checked by the reviewers before merging the pull request:_**

As a reviewer please read through all the code lines and make sure that the code is fully understood, bug free, well-documented and well-structured.

#### General
- [x] The reviewer executed the new code features at least once and checked the results manually

- [x] The code follows the [t8code coding guidelines](https://github.com/holke/t8code/wiki/Coding-Guideline)
- [x] New source/header files are properly added to the Makefiles
- [x] The code is well documented
- [x] All function declarations, structs/classes and their members have a proper doxygen documentation
- [x] All new algorithms and data structures are sufficiently optimal in terms of memory and runtime (If this should be merged, but there is still potential for optimization, create a new issue)

#### Tests
- [x] The code is covered in an existing or new test case using Google Test

#### Github action

- [x] The code compiles without warning in debugging and release mode, with and without MPI (this should be executed automatically in a github action)
- [x] All tests pass (in various configurations, this should be executed automatically in a github action)

  If the Pull request introduces code that is not covered by the github action (for example coupling with a new library):
  - [ ] Should this use case be added to the github action?
  - [ ] If not, does the specific use case compile and all tests pass (check manually)

#### Scripts and Wiki

- [ ] If a new directory with source-files is added, it must be covered by the `script/find_all_source_files.scp` to check the indentation of these files.
- [ ] If this PR introduces a new feature, it must be covered in an example/tutorial and a Wiki article.

#### Licence

- [x] The author added a BSD statement to `doc/` (or already has one)

### CODE REVIEW

File: test/t8_schemes/t8_gtest_scheme_helper.hxx [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+  {
+    int level = ts->t8_element_level (element);
+    ASSERT_LE (level, maxlvl);
+    ASSERT_LT (maxlvl, ts->t8_element_maxlevel ());
+
+    check_element (element);
+
+    if (ts->t8_element_level (element) < maxlvl) {
+      /* iterate over all children */
+      int num_children = ts->t8_element_num_children (element);
```
Comment (Davknapp) [inline test/t8_schemes/t8_gtest_scheme_helper.hxx:51]:
```suggestion
      const int num_children = ts->t8_element_num_children (element);
```

File: test/t8_schemes/t8_gtest_scheme_helper.hxx [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+  {
+    dfs_test_teardown();
+  }
+
+  t8_scheme_cxx *scheme;
+  t8_eclass_t eclass;
+  t8_eclass_scheme_c *ts;
+  t8_element_t *element;
+
+  const int maxlvl = 4;
```
Comment (Davknapp) [inline test/t8_schemes/t8_gtest_scheme_helper.hxx:92]: Please remove it from the class and make it a parameter of check_recursive_dfs

File: test/t8_schemes/t8_gtest_scheme_helper.hxx [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+*/
+
+#ifndef T8_GTEST_SCHEME_HELPER_H
+#define T8_GTEST_SCHEME_HELPER_H
+
+#include <gtest/gtest.h>
+#include <t8_eclass.h>
+#include <t8_schemes/t8_default/t8_default_cxx.hxx>
+
+#define AllImplementations testing::Range (T8_ECLASS_ZERO, T8_ECLASS_COUNT)
```
Comment (Davknapp) [inline test/t8_schemes/t8_gtest_scheme_helper.hxx:30]: This should be moved into a seperate header, as it has nothing to do with the dfs itself. Then you could also rename the file into t8_gtest_dfs_blueprint.hxx?

**Review by Davknapp — CHANGES_REQUESTED** [review]
Nice addition!

File: test/t8_gtest_macros.hxx [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with t8code; if not, write to the Free Software Foundation, Inc.,
+51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+*/
+
+/** \file t8_gtest_custom_assertion.cxx
```
Comment (Davknapp) [inline test/t8_gtest_macros.hxx:23]: Update this comment to match the file

**Review by Davknapp — CHANGES_REQUESTED** [review]

File: test/t8_gtest_macros.hxx [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+* Provide customized GoogleTest functions for improved error-output
+*/
```
Comment (lukasdreyer) [inline test/t8_gtest_macros.hxx:25]:
```suggestion
/** \file t8_gtest_macros.hxx
* Provide macros for instantiating parameterized tests
*/
```

**Review by Davknapp — APPROVED** [review]

**Pull request merged**